src/commands/plugins.rs

```diff
@@ -0,0 +1,275 @@
+    pub downgrade: bool,
+}
+
+impl Upgrade {
+    /// Upgrades one or all plugins by reinstalling the latest or a specified
+    /// version of a plugin. If downgrade is specified, first uninstalls the
+    /// plugin.
+    pub async fn run(self) -> Result<()> {
+        let plugins_dir = get_spin_plugins_directory()?;
+        let spin_version = env!("VERGEN_BUILD_SEMVER");
+        let manifest_dir = plugins_dir.join(PLUGIN_MANIFESTS_DIRECTORY_NAME);
+
+        // Check if no plugins are currently installed
+        if !manifest_dir.exists() {
+            println!("No currently installed plugins to update.");
+            return Ok(());
+        }
+
+        if self.all {
+            // Install the latest of all currently installed plugins
+            for plugin in std::fs::read_dir(manifest_dir)? {
+                let path = plugin?.path();
+                let name = path
+                    .file_stem()
+                    .ok_or_else(|| anyhow!("expected directory for plugin"))?
+                    .to_str()
+                    .ok_or_else(|| anyhow!("Cannot convert directory to String"))?
+                    .to_string();
+                if let Err(e) = PluginInstaller::new(
+                    ManifestLocation::PluginsRepository(PluginInfo::new(
+                        &name,
+                        Url::parse(SPIN_PLUGINS_REPO)?,
+                        None,
+                    )),
+                    plugins_dir.clone(),
+                    self.yes_to_all,
+                    spin_version,
+                )
+                .install()
+                .await
+                {
+                    // Ignore plugins that were not installed from the central
+                    // plugins repository
+                    if e.to_string().contains("Could not find plugin") {
+                        log::info!(
+                            "Could not update {} plugin as DNE in central repository",
+                            name
+                        );
+                    } else {
+                        return Err(e);
+                    }
+                }
+            }
+        } else {
+            let name = self
+                .name
+                .ok_or_else(|| anyhow!("plugin name is required for upgrades"))?;
+            // If downgrade is allowed, first uninstall the plugin
+            if self.downgrade {
+                PluginUninstaller::new(&name, plugins_dir.clone()).run()?;
+            }
+            let manifest_location = match (self.local_manifest_src, self.remote_manifest_src) {
+                (Some(path), None) => ManifestLocation::Local(path),
+                (None, Some(url)) => ManifestLocation::Remote(url),
+                _ => ManifestLocation::PluginsRepository(PluginInfo::new(
+                    &name,
+                    Url::parse(SPIN_PLUGINS_REPO)?,
+                    self.version,
+                )),
+            };
+            PluginInstaller::new(
```

Comment:
This is very subjective, but my eyes keep speed-bumping on the `Foo::new(several-lines-of-stuff).operation()` pattern.  I want `new` to return a value and keep being surprised that it doesn't, particularly when the "oh it's just to call the method" realisation is so far away.

If we really aren't interested in these values as stateful, how about having a `PluginInstaller::install(...)` function for when you have no interest in keeping the installer object around?  (Just wrapping the `new/install` combo.)

Reply:
got rid of the `PluginInstaller` in favor of a `PluginManager` with `install` and `uninstall` functions. Thank you for pointing this out!